### ISSUE TITLE
[HOLD for payment 2024-09-17] [$250] Add Offline and Error Patterns to Remaining Integrations

### ISSUE THREAD

**Issue by yuwenmemon** [Expensify/App#45406]
With https://github.com/Expensify/App/pull/44739 we added offline and error patterns to the Sage Intacct connection menu.

One observation from discussion in [Slack](https://expensify.slack.com/archives/C06ML6X0W9L/p1720722381758239?thread_ts=1720475656.078959&cid=C06ML6X0W9L) was that this pattern is not necessarily implemented for the other integrations (QBO, Xero, and NetSuite) and we'd like to maintain parity between all integrations menus. 

### Solution
Implement the same offline and error patterns implemented for Intacct in the other integrations

cc @war-in @JmillsExpensify @shawnborton @trjExpensify 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014b0b4b88631ddf1d</li>
        <li>Upwork Job ID: 1812869868565563336</li>
        <li>Last Price Increase: 2024-07-15</li>
<li>Automatic offers: </li>
<ul>
<li>ikevin127 | Reviewer | 103126194</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @stephanieelliott</details>

**Comment by war-in** [comment]
Hi, I'm Marcin from Software Mansion and I'd like to work on this issue!

**Comment by ikevin127** [comment]
@kabeer95 Thanks for your proposal. Unfortunately this issue was already assigned to Marcin from Software Mansion based on https://github.com/Expensify/App/issues/45406#issuecomment-2228787133. Additionally, in the future - in order for reviewers to be able to consider your proposal you need to precisely follow the [proposal template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md) when posting a proposal.

**Comment by kabeer95** [comment]
@ikevin127 okay.

**Comment by war-in** [comment]
Posting my update because I'll be OOO till Monday

I started working on fixing offline and error patterns in Xero integration and created a [draft PR](https://github.com/Expensify/App/pull/45493). You can check the web video in the PR description but I haven't worked on offline handling yet.

The best (and fastest) way will be splitting this work into a few PRs (one per integration). It should help with testing.

**Comment by war-in** [comment]
I also found and fixed some bugs (e.g. showing offline state for all export subpages when only one is updated). It will require thorough review but @hungvu193 is up-to-date (he reviewed my Sage Intacct offline and error PR) so we should be able to merge things fast 🚀

**Comment by ikevin127** [comment]
Will @hungvu193 review all PRs here or just 1 of them ? Asking to know whether I should unassign myself.

**Comment by war-in** [comment]
@hungvu193 What are your thoughts?

**Comment by hungvu193** [comment]
@ikevin127  Hey, this issue is not urgent so if you're interested, please go ahead Kev 😄

**Comment by ikevin127** [comment]
Not overdue, **war-in** is working on PRs.

**Comment by war-in** [comment]
> Not overdue, **war-in** is working on PRs.

That's right! The [first one](https://github.com/Expensify/App/pull/45493) (Xero) is ready for internal review :)

**Comment by war-in** [comment]
I'm working on the Sage Intacct integration now and there are some places where two components can modify the same onyx value. If the request fails we should somehow show the error message but should we show it only in one place or maybe both?

Example 

https://github.com/user-attachments/assets/4544296b-1d04-4d20-a245-28d0bb26bbdd

- (0:01): toggle sets some key
- (0:03): request fails and error is there (so far so good)
- (0:04): toggle sets key again and menu item is shown
- (0:05): error appears in both toggle and menu item
- (0:09): we clear the error and it disappears from both places

Is this behaviour correct? Or should we show this error only inside the menu item? 🤔 

In my opinion, the best solution is to show an error message below the disabled toggle and not to show it when the toggle is enabled

https://github.com/user-attachments/assets/e307df42-c807-4c03-8bbb-09da3acd1dbc

cc @shawnborton

**Comment by zfurtak** [comment]
hi there, I'm from SWM and would like to help @war-in with this issue :) I will take care of QBO part 👩🏽‍💻

**Comment by shawnborton** [comment]
Hmm @Expensify/design @yuwenmemon  any strong feelings about the comment [above](https://github.com/Expensify/App/issues/45406#issuecomment-2252788844)? I don't have a strong opinion really.

**Comment by dubielzyk-expensify** [comment]
How does one fix this error? Is it just a connection issue that'll resolve by itself?

I generally prefer one error unless there's multiple places to fix it, but also don't feel too strongly.

**Comment by war-in** [comment]
> Is it just a connection issue that'll resolve by itself?

The error shows when the API call fails. To fix it someone will have to choose the value again

**Comment by dubielzyk-expensify** [comment]
Then perhaps show it both places given one is a toggle 👍

**Comment by war-in** [comment]
This issue should not be closed until we handle all integrations. We finished `Xero` and have a PR for `Sage Intacct`. `QBO` is in development by @zfurtak and one of us will take `NetSuite` in the upcoming days

**Comment by ikevin127** [comment]
* [x] [@ikevin127] Please propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the new feature will work correctly on production in further releases.

## Regression Test Proposal - https://github.com/Expensify/App/pull/45493

**A. Error testing**

1. Set `Simulate failing network` in `Troubleshooting` (Ctrl / CMD + D).
2. Go to `Settings` -> `Workspaces` -> your Xero workspace -> `Accounting`.
3. Go through every page and try changing values.
4. When the requests fail, verify that  RBRs and error messages are displayed.

**B. Offline testing**

1. Go to `Settings` -> `Workspaces` -> your Xero workspace -> `Accounting`.
2. Set `Force offline` in `Troubleshooting` (Ctrl / CMD + D).
3. Go through every page and try changing values.
4. Verify that changed values / fields are displayed with 50% opacity.

Do we agree 👍 or 👎.

**Comment by war-in** [comment]
Hi all 👋 I'm currently working on the `NetSuite` integration update (offline and error patterns) and I saw that on every API call failure, we revert the state to the previous value.

That's not the case in any other integration and OldDot (checked with Xero). 

I think we should update this behaviour for the NetSuite integration to align it with the others.
I can handle it here or create a separate PR (the second option is preferred because my NetSuite update PR is already big).

cc @trjExpensify

**Comment by trjExpensify** [comment]
Yeah, we apply offline pattern B here right? 

So if it fails, we show the in-line error message on the updated value and then when you dismiss we revert to the previous value.

**Comment by trjExpensify** [comment]
Ah gotcha, I think that's the standard for pattern B across the board in App for failed add/updated values.

**Comment by war-in** [comment]
> Ah gotcha, I think that's the standard for pattern B across the board in App for failed add/updated values.

That's clear now :) Let's use this issue to keep track of implementing this pattern correctly in all integrations.
All of them will require additional work

**Comment by war-in** [comment]
@trjExpensify This is what I found:

- we would have to store failure data in Onyx to be able to access them while closing errors
- those failure data would bring back the previous state (from before the changes)
- to achieve that we need to create a new Onyx key under the `policy_[someID].connections.[someConnection]` with the same structure as a connection. Let's call it `failureData` (we could access it like that: `policy_[someID].connections.[someConnection].failureData`)

Let me know what you think about all this and if we want to put some effort here I'll prepare a POC (proposal) 🙏

**Comment by war-in** [comment]
A much easier approach is one currently used in `NetSuite` integration -> removing changed value and displaying an error.

But it has drawbacks too. It's not always obvious where to show the error because sometimes optimistic data adds a row and if this row disappears there is no correct place to show it.

Considering the above I think we should focus on adding failure data (maybe the right name is previous data 👀 ) to Onyx

**Comment by ikevin127** [comment]
**Notes:**
- [comment](https://github.com/Expensify/App/issues/45406#issuecomment-2256634623) on the first PR (which requires payment today), and my BZ Checklist [here](https://github.com/Expensify/App/issues/45406#issuecomment-2264140526)
- [comment](https://github.com/Expensify/App/issues/45406#issuecomment-2257911412) regarding why this issue shouldn't be closed after the first payment

cc @stephanieelliott

**Comment by 289Adam289** [comment]
Hi! I'm from SWM and I would like to help @war-in with Sage Intacct part of this issue.

**Comment by war-in** [comment]
@289Adam289 will update the behaviour on failing network requests discussed [here](https://swmansion.slack.com/archives/C036QM0SLJK/p1722855119292609).

tl;dr We want to revert optimistic data on failed requests

**Comment by ikevin127** [comment]
cc @stephanieelliott 👋 Bump for the first payment which was due **yesterday** (see https://github.com/Expensify/App/issues/45406#issuecomment-2270057330).

**Comment by ikevin127** [comment]
cc @stephanieelliott 👋 Bump for the first payment which was due **yesterday** (see https://github.com/Expensify/App/issues/45406#issuecomment-2270057330).

**Comment by stephanieelliott** [comment]
Sorry, I think I'm confused about the payment situation -- are we paying in installments here? Generally 1 issue = $250, irrespective of the number of PRs, paid when final PR is merged +7 (which it seems like is 8/13). Standalone PRs typically get their own auto-created payment issue. 

I've paid out the existing offer, but I'm not sure how to handle future payments. Left the Upwork contract open in case we need to create multiple milestones, I'll just need to know how many.

**Comment by ikevin127** [comment]
ℹ️ Notice that @shubham1206agra will take over the NetSuite PR https://github.com/Expensify/App/pull/46667 as they have credentials for the integration and I don't - the passing over was [discussed](https://expensify.slack.com/archives/C02NK2DQWUX/p1723024882613169) on Slack. Please remove me from the PR as reviewer and assign them!

Also @hungvu193 took over the SageIntaact PR https://github.com/Expensify/App/pull/46294 for the same reason. This PR is already merged and payment is coming up on 13th.

@stephanieelliott Both C+ reviewers mentioned above will require payment as part of this issue.

##

cc @yuwenmemon @deetergp 


## 

I'm continuing with the review of the other PRs / follow-ups (QBO and Xero).

**Comment by ikevin127** [comment]
What's the difference between:
- https://github.com/Expensify/App/pull/46294
    and
- https://github.com/Expensify/App/pull/47079

I thought we already handled that one 🤔

cc @war-in @289Adam289

**Comment by hungvu193** [comment]
Yeah, I can take that one 😄 . Hey @war-in @289Adam289, can you answer the above question?

**Comment by hungvu193** [comment]
I see, I'll review today 😄

**Comment by stephanieelliott** [comment]
Ok, so I want to make sure i have this right -- these are the PRs and reviewers due payment I have on this issue @hungvu193 @ikevin127 can you look this over and confirm I have the relevant PRs and correct person due payment? 

 - **Sage** 
   - https://github.com/Expensify/App/pull/44739 - reviewer @ikevin127  [PAID 8/5]
   - https://github.com/Expensify/App/pull/46294 - reviewer @hungvu193 [DUE 8/13] 

 - **NetSuite**  
   -  https://github.com/Expensify/App/pull/46667 - reviewer @hungvu193

 - **QBO:** 
   - https://github.com/Expensify/App/pull/46511 - reviewer @ikevin127

 - **Xero:** 
   - https://github.com/Expensify/App/pull/45493 - @ikevin127 [UNPAID, was due 8/5]

**Comment by hungvu193** [comment]
I reviewed 2 SageIntacct PRs, but no payment needed for me, I will be paid centrally later when SageIntacct project is completed

**Comment by ikevin127** [comment]
@stephanieelliott The 2 Sage Intaact merged PRs were both reviewed by @hungvu193 which as they mentioned will not be due payment per every PR.

The Xero PR I reviewed first was already paid out to me on 8/5.

##

So far we're all good with payments, there will be other PRs reviewed by me as part of this issue where I'll need payment once the regression period is over.

**Comment by ikevin127** [comment]
⚠️ Production deploy automation failed here -> a new payment date should be on **[HOLD for Payment 2024-08-22]** according to yesterday’s production deploy confirmed in https://github.com/Expensify/App/pull/46898#issuecomment-2294134343 and [deploy checklist](https://github.com/Expensify/App/issues/47356).

I will require payment as reviewer of PR:
- https://github.com/Expensify/App/pull/46898

cc @stephanieelliott

**Comment by ikevin127** [comment]
cc @stephanieelliott

**Comment by stephanieelliott** [comment]
Summarizing payment for https://github.com/Expensify/App/pull/46898

Contributor @war-in - no payment due (agency)
C+ @ikevin127 -$250 via Upwork on [this job](https://www.upwork.com/jobs/~014b0b4b88631ddf1d), milestone 2 - PAID

**Comment by stephanieelliott** [comment]
Status check on the remaining PRs on this issue: 

   - https://github.com/Expensify/App/pull/46667 - reviewer @ikevin127, `Reviewing`
   - https://github.com/Expensify/App/pull/46511 - reviewer @ikevin127, `Reviewing`

**Comment by ikevin127** [comment]
The NetSuite one was taken by @shubham1206agra as C+ reviewer since only they have the credentials required to trst NetSuite integration.

The 2nd one is correct, currently waiting kn CME for final approve and merge, hopefully soon 😁

**Comment by stephanieelliott** [comment]
> https://github.com/Expensify/App/pull/46667 - reviewer @shubham1206agra, Reviewing
> https://github.com/Expensify/App/pull/46511 - reviewer @ikevin127, Reviewing

Both on staging!

**Comment by ikevin127** [comment]
⚠️ Automation failed here -> the last 2 payments should be on **[HOLD for Payment 2024-09-17]** according to 3 days ago production deploy from https://github.com/Expensify/App/issues/48664#issuecomment-2338463845 (both PRs are in this deploy checklist, first two).

**Notes for payments**
- ⚠️ looks like the NetSuite one had a regression and the QBO one had no issues

cc @stephanieelliott

**Comment by ikevin127** [comment]
cc @stephanieelliott

**Comment by stephanieelliott** [comment]
All paid! Thanks
